pinot-controller/src/main/java/org/apache/pinot/controller/cursors/ResponseStoreCleaner.java

```diff
@@ -0,0 +1,222 @@
+import org.apache.pinot.controller.helix.core.periodictask.ControllerPeriodicTask;
+import org.apache.pinot.spi.auth.AuthProvider;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * ResponseStoreCleaner periodically gets all responses stored in a response store and deletes the ones that have
+ * expired. From each broker, tt gets the list of responses. Each of the response has an expiration unix timestamp.
+ * If the current timestamp is greater, it calls a DELETE API for every response that has expired.
+ */
+public class ResponseStoreCleaner extends ControllerPeriodicTask<Void> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResponseStoreCleaner.class);
+  private static final int TIMEOUT_MS = 3000;
+  private static final String QUERY_RESULT_STORE = "%s://%s:%d/responseStore";
+  private static final String DELETE_QUERY_RESULT = "%s://%s:%d/responseStore/%s";
+  // Used in tests to trigger the delete instead of waiting for the wall clock to move to an appropriate time.
+  public static final String CLEAN_AT_TIME = "response.store.cleaner.clean.at.ms";
+  private final ControllerConf _controllerConf;
+  private final Executor _executor;
+  private final PoolingHttpClientConnectionManager _connectionManager;
+  private final AuthProvider _authProvider;
+
+  public ResponseStoreCleaner(ControllerConf config, PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, ControllerMetrics controllerMetrics, Executor executor,
+      PoolingHttpClientConnectionManager connectionManager) {
+    super("ResponseStoreCleaner", getFrequencyInSeconds(config), getInitialDelayInSeconds(config),
+        pinotHelixResourceManager, leadControllerManager, controllerMetrics);
+    _controllerConf = config;
+    _executor = executor;
+    _connectionManager = connectionManager;
+    _authProvider =
+        AuthProviderUtils.extractAuthProvider(config, CommonConstants.CursorConfigs.RESPONSE_STORE_AUTH_PREFIX);
+  }
+
+  private static long getInitialDelayInSeconds(ControllerConf config) {
+    long initialDelay = config.getPeriodicTaskInitialDelayInSeconds();
+    String responseStoreCleanerTaskInitialDelay =
+        config.getProperty(CommonConstants.CursorConfigs.RESPONSE_STORE_CLEANER_INITIAL_DELAY);
+    if (responseStoreCleanerTaskInitialDelay != null) {
+      initialDelay = TimeUnit.SECONDS.convert(TimeUtils.convertPeriodToMillis(responseStoreCleanerTaskInitialDelay),
+          TimeUnit.MILLISECONDS);
+    }
+    return initialDelay;
+  }
+
+  private static long getFrequencyInSeconds(ControllerConf config) {
+    long frequencyInSeconds = TimeUnit.SECONDS.convert(
+        TimeUtils.convertPeriodToMillis(CommonConstants.CursorConfigs.DEFAULT_RESPONSE_STORE_CLEANER_FREQUENCY_PERIOD),
+        TimeUnit.MILLISECONDS);
```

Comment:
We could do this in an else block below instead to avoid unnecessary conversion when the `controller.cluster.response.store.cleaner.frequencyPeriod` config is set.

Reply:
I agree if it's directly using the default value, but here we're also doing string parsing and time unit conversions. This is only a nitpick though as it isn't in a particularly performance sensitive path.

Reply:
Also the default will be used super-majority of the times.